src/execution_plan/execution_plan.c

```diff
@@ -32,11 +32,41 @@ static inline void _ExecutionPlan_UpdateRoot(ExecutionPlan *plan, OpBase *new_ro
 	plan->root = new_root;
 }
 
+// For all ops that refer to QG entities, rebind them with the matching entity
+// in the provided QueryGraph.
+// (This logic is ugly, but currently necessary.)
+static void _RebindQueryGraphReferences(OpBase *op, const QueryGraph *qg) {
+	switch(op->type) {
+	case OPType_INDEX_SCAN:
+		((IndexScan *)op)->n = QueryGraph_GetNodeByAlias(qg, ((IndexScan *)op)->n->alias);
+		return;
+	case OPType_ALL_NODE_SCAN:
+		((AllNodeScan *)op)->n = QueryGraph_GetNodeByAlias(qg, ((AllNodeScan *)op)->n->alias);
+		return;
+	case OPType_NODE_BY_LABEL_SCAN:
+		((NodeByLabelScan *)op)->n = QueryGraph_GetNodeByAlias(qg, ((NodeByLabelScan *)op)->n->alias);
+		return;
+	case OPType_EXPAND_INTO:
+		AlgebraicExpression_UpdateReferences(qg, ((OpExpandInto *)op)->ae);
+		return;
+	case OPType_CONDITIONAL_TRAVERSE:
+		AlgebraicExpression_UpdateReferences(qg, ((CondTraverse *)op)->ae);
+		return;
+	case OPType_CONDITIONAL_VAR_LEN_TRAVERSE:
+	case OPType_CONDITIONAL_VAR_LEN_TRAVERSE_EXPAND_INTO:
+		AlgebraicExpression_UpdateReferences(qg, ((CondVarLenTraverse *)op)->ae);
+		return;
+	default:
+		return;
+	}
+}
```

Comment:
I believe that if we move from holding a reference to a QueryGraph objects either nodes or edges to aliases then this logic can be removed, see my comment on AlgebraicExpression